src/core/lib/event_engine/workqueue.h

```diff
@@ -0,0 +1,238 @@
+  }
+
+ private:
+  // Attempts to pop from the front of the queue (oldest).
+  // This will return nullopt if the queue is empty, or if other workers
+  // are already attempting to pop from this queue.
+  absl::optional<T> TryLockAndPop(bool front) ABSL_LOCKS_EXCLUDED(mu_) {
+    // Do not block the worker if there are other workers trying to pop
+    // tasks from this queue.
+    if (!mu_.TryLock()) return absl::nullopt;
+    auto mu_cleanup = absl::MakeCleanup([this]() {
+      mu_.AssertHeld();
+      mu_.Unlock();
+    });
+    if (GPR_UNLIKELY(elements_.empty())) {
+      if (most_recent_element_enqueue_timestamp_.load(
+              std::memory_order_relaxed) == kInvalidTimestamp) {
+        return absl::nullopt;
+      }
+      if (!most_recent_element_lock_.TryLock()) return absl::nullopt;
+      absl::optional<T> ret = absl::nullopt;
+      if (GPR_LIKELY(most_recent_element_.has_value())) {
+        most_recent_element_enqueue_timestamp_.store(kInvalidTimestamp,
+                                                     std::memory_order_relaxed);
+        ret = std::exchange(most_recent_element_, absl::nullopt);
+      }
+      most_recent_element_lock_.Unlock();
+      return ret;
+    }
+    // the queue has elements, let's pop one and update timestamps
+    Storage ret_s;
+    if (front) {
+      ret_s = std::move(elements_.front());
+      elements_.pop_front();
+    } else {
+      ret_s = std::move(elements_.back());
+      elements_.pop_back();
+    }
+    if (elements_.empty()) {
+      oldest_enqueued_timestamp_.store(kInvalidTimestamp,
+                                       std::memory_order_relaxed);
+    } else if (front) {
+      oldest_enqueued_timestamp_.store(elements_.front().enqueued(),
+                                       std::memory_order_relaxed);
+    }
+    return ret_s.TakeElement();
+  }
+
+  // Attempts to pop from the back of the queue (most recent).
+  // This will return nullopt if the queue is empty, or if other workers
+  // are already attempting to pop from this queue.
+  absl::optional<T> TryPopMostRecentElement() {
+    if (!most_recent_element_lock_.TryLock()) return absl::nullopt;
+    if (GPR_UNLIKELY(!most_recent_element_.has_value())) {
+      most_recent_element_lock_.Unlock();
+      return absl::nullopt;
+    }
+    most_recent_element_enqueue_timestamp_.store(kInvalidTimestamp,
+                                                 std::memory_order_relaxed);
+    absl::optional<T> tmp = std::exchange(most_recent_element_, absl::nullopt);
+    most_recent_element_lock_.Unlock();
+    return tmp;
+  }
+
+  // The managed items in the queue
+  std::deque<Storage> elements_ ABSL_GUARDED_BY(mu_);
+  // The most recently enqueued element. This is reserved from work stealing
+  absl::optional<T> most_recent_element_
+      ABSL_GUARDED_BY(most_recent_element_lock_);
+  absl::base_internal::SpinLock ABSL_ACQUIRED_AFTER(mu_)
```

Comment:
I'm moderately worried about `SpinLock` safety outside of kernel schedulers we control - especially since there's a Lock() (and not TryLock()) in Add.

@soheilhy what's your intuition about how necessary this is, and could we perhaps just use a `Mutex` here?

Reply:
(I'm also quite interested in @soheilhy 's take, but to add ...)

This was a first pass, and I wanted to discuss it. I'm getting some benchmarks set up, and I planned to swap this with a Mutex to see the effect. I'd prefer not to rely on anything in absl's `base_internal` namespace as well, but I think we can consider tolerating it if the performance benefit is significant. We also have gpr_spinlock, which I could wrap in something that plays well with absl lockable types.

Reply:
This used to be an atomic, but after adding timestamps we ended up adding a spinlock. It really depends on how this is used. You can start with `Mutex` and switch back to `SpinLock` later.

Reply:
Some quick benchmark results on my machine using a Mutex instead of a Spinlock:

single-element tests: 66% decrease in performance (~3.6M/s to ~1.2M/s)
multiple-element tests: 22% decrease in performance (~1.1M/s to 860k/s)

```
------------------------------------------------------------------------------------------
Benchmark                                Time             CPU   Iterations UserCounters...
------------------------------------------------------------------------------------------
BM_WorkQueueIntptrPopFront/1           807 ns          806 ns       865773 items_per_second=1.24034M/s
BM_WorkQueueIntptrPopFront/8          8991 ns         8989 ns        75894 items_per_second=889.984k/s
BM_WorkQueueIntptrPopFront/64        75695 ns        75674 ns         9436 items_per_second=845.735k/s
BM_WorkQueueIntptrPopFront/512      589618 ns       589446 ns         1162 items_per_second=868.612k/s
BM_WorkQueueIntptrPopFront/4096    4857182 ns      4855613 ns          145 items_per_second=843.56k/s
I0902 19:37:09.771871994      12 test_config.cc:194]         TestEnvironment ends
================================================================================
```

Reply:
correction: this was a debug build. There is effectively no difference in opt builds, ~12M/s for all sizes > 1. I'll write some multithreaded microbenchmarks.

Reply:
Performance update below. Spinlock has been removed.